voice/src/main/java/discord4j/voice/VoiceGatewayClient.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Discord4J is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Discord4J is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package discord4j.voice;
+
+import com.darichey.simplefsm.FiniteStateMachine;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.iwebpp.crypto.TweetNaclFast;
+import discord4j.voice.VoiceGatewayEvent.Start;
+import discord4j.voice.VoiceGatewayEvent.Stop;
+import discord4j.voice.VoiceGatewayState.*;
+import discord4j.voice.json.*;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
+import io.netty.handler.codec.http.websocketx.WebSocketFrame;
+import reactor.core.Disposable;
+import reactor.core.publisher.EmitterProcessor;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.netty.NettyPipeline;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.websocket.WebsocketInbound;
+import reactor.netty.http.websocket.WebsocketOutbound;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+import java.io.InputStream;
+import java.time.Duration;
+
+import static io.netty.handler.codec.http.HttpHeaderNames.USER_AGENT;
+
+public class VoiceGatewayClient {
+
+    private final Logger log = Loggers.getLogger("discord4j.voice.gateway.client");
+    private final FiniteStateMachine<VoiceGatewayState, VoiceGatewayEvent> gatewayFSM;
+
+    private final EmitterProcessor<VoiceGatewayPayload<?>> sender = EmitterProcessor.create(false);
+    private final ObjectMapper mapper;
+    final VoiceSocket voiceSocket;
+
+    public VoiceGatewayClient(long serverId, long userId, String sessionId, String token, ObjectMapper mapper,
+                              Scheduler scheduler, AudioProvider provider, AudioReceiver receiver) {
+        this.mapper = mapper;
+        this.voiceSocket = new VoiceSocket();
+        this.gatewayFSM = new FiniteStateMachine<VoiceGatewayState, VoiceGatewayEvent>() {{
+            startWith(Stopped.INSTANCE);
+
+            when(Stopped.class)
+                    .on(Start.class, (curState, start) -> {
+                        Disposable websocketTask = HttpClient.create()
+                                .wiretap(true)
+                                .headers(headers -> headers.add(USER_AGENT, "DiscordBot(https://discord4j.com, 3)"))
```

Comment:
This user-agent should probably stored in a constant in common and/or should be configurable

Reply:
If we want to change it, it should be done in conjunction with GatewayClient.

Reply:
Would be appropriate on a separate PR/commit

Reply:
A TODO tag would be appreciated here